projects/odf_transform/odf_transform/odf_parser/odf_parser.py

```diff
@@ -207,54 +207,91 @@ def _find_previous_key(key_list, key):
         return previous_key
 
     # Find matching vocabulary
-    if type(vocabulary) is dict:
-        flag_dict = {}
-        for var in metadata.keys():
-            # Retrieve ODF CODE and Associated number
-            odf_parameter_code = metadata[var][parameter_code_att]
-            # Separate the parameter_code from the number at the end of the variable
-            parameter_code = odf_parameter_code.rsplit('_', 1)
-
-            # Retrieve trailing
-            # Parameter codes are generally associated with a trailing number the define the
-            # primary, secondary ,... data
-            metadata[var]['parameter_code'] = parameter_code[0]
-            if len(parameter_code) == 2 and parameter_code[1] != '':
-                metadata[var]['parameter_code_number'] = int(parameter_code[1])
+
+    flag_dict = {}
+    for var in metadata.keys():
+        # Retrieve ODF CODE and Associated number
+        odf_parameter_code = metadata[var][parameter_code_att]
+        # Separate the parameter_code from the number at the end of the variable
+        parameter_code = odf_parameter_code.rsplit('_', 1)
+
+        # Retrieve trailing
+        # Parameter codes are generally associated with a trailing number the define the
+        # primary, secondary ,... data
+        metadata[var]['parameter_code'] = parameter_code[0]
+        if len(parameter_code) == 2 and parameter_code[1] != '':
+            metadata[var]['parameter_code_number'] = int(parameter_code[1])
+        else:
+            metadata[var]['parameter_code_number'] = 1
+
+        # FLAG VARIABLES Detect if it is a flag column associated with another column
+        flag_column = False
+        if parameter_code[0].startswith('QQQQ'):  # MLI FLAG should apply to previous variable
+            flag_dict[odf_parameter_code] = _find_previous_key(metadata, var)
+            flag_column = True
+        elif parameter_code[0] in ['QCFF', 'FFFF']:
+            flag_column = True
+        elif parameter_code[0].startswith('Q') and odf_parameter_code[1:] in metadata.keys():
+            # BIO Format which Q+[PCODE] of the associated variable
+            flag_dict[odf_parameter_code] = odf_parameter_code[1:]
+            flag_column = True
+        # Make sure that the flag column relate to something
+        if flag_column and parameter_code[0] not in ['QCFF', 'FFFF'] and flag_dict[odf_parameter_code] not in metadata:
+            warnings.warn(odf_parameter_code + ' flag is refering to' + \
+                          flag_dict[odf_parameter_code] + ' which is not available as variable',
+                          UserWarning)
+
+        # Loop through each organisations and find the matching parameter_code within the vocabulary
+        standardized_variables = False
+        if type(vocabulary) is pd.DataFrame and not flag_column and var not in ['SYTM_01']:
```

Comment:
which other type are you expecting `vocabulary` could be?

Reply:
yeah we'll move away from the json ok I'll add those changes

Reply:
it was replace by `vocabulary is not None`